_includes/exhibits.js

```diff
@@ -0,0 +1,1929 @@
+  return keys.reduce(function(h, k) {
+    var value = state[k] || 0;
+    // Array separated by underscore
+    if (value.constructor === Array) {
+      value = value.join('_');
+    }
+    return h + delimit + k + '=' + value;
+  }, '').slice(1);
+
+};
+
+const deserialize = function(entries) {
+  const query = entries.reduce(function(o, entry) {
+    if (entry) {
+      const kv = entry.split('=');
+      const val = kv.slice(1).join('=') || '1';
+      const vals = val.split('_');
+      const key = kv[0];
+      // Handle arrays or scalars
+      o[key] = vals.length > 1? vals: val;
+    }
+    return o;
+  }, {});
+
+  return query;
+};
+
+const newCopyYamlButton = function(THIS) {
+  const copy_pre = 'Copy to Clipboard';
+  const copy_post = 'Copied';
+  $(this).tooltip({
+    title: copy_pre
+  });
+
+  $(this).on('relabel', function(event, message) {
+    $(this).attr('data-original-title', message).tooltip('show');
+  });
+
+  $(this).click(function() {
+    $(this).trigger('relabel', [copy_post]);
+    ctrlC(THIS.bufferYaml);
+    setTimeout(function() {
+      $(this).trigger('relabel', [copy_pre]);
+    }, 1000);
+    return false;
+  });
+};
+
+const newCopyButton = function() {
+  const copy_pre = 'Copy to Clipboard';
+  const copy_post = 'Copied';
+  $(this).tooltip({
+    title: copy_pre
+  });
+
+  $(this).on('relabel', function(event, message) {
+    $(this).attr('data-original-title', message).tooltip('show');
+  });
+
+  $(this).on('click', function() {
+    const form = $(this).closest('form');
+    const formData = parseForm(form);
+    $(this).trigger('relabel', [copy_post]);
+    ctrlC(formData.copy_content);
+    setTimeout(function() {
+      $(this).trigger('relabel', [copy_pre]);
+    }, 1000);
+  return false;
+  });
+};
+
+const changeSprings = function(viewer, seconds, stiffness) {
+  const springs = [
+    'centerSpringX', 'centerSpringY', 'zoomSpring'
+  ];
+  springs.forEach(function(spring) {
+    const s = viewer.viewport[spring];
+    s.animationTime = seconds;
+    s.springStiffness = stiffness;
+    s.springTo(s.target.value);
+  });
+};
+
+const HashState = function(viewer, tileSources, exhibit, options) {
+
+  this.resetCount = 0;
+  this.embedded = options.embedded || false;
+  this.showdown = new showdown.Converter();
+  this.tileSources = tileSources;
+  this.exhibit = exhibit;
+  this.viewer = viewer;
+  viewer.setVisible(false);
+
+  this.hashable = {
+    exhibit: [
+      's', 'w', 'g', 'v'
+    ],
+    edits: [
+      's', 'w', 'g', 'v', 'o'
+    ],
+    tag: [
+      'd', 'o', 'g', 'v'
+    ]
+  };
+  this.searchable = {
+    ui: [
+      'edit'
+    ],
+  };
+
+  this.state = {
+    buffer: {
+      waypoint: undefined
+    },
+    changed: false,
+    design: {},
+    w: [0],
+    g: 0,
+    s: 0,
+    v: [1, 0.5, 0.5],
+    o: [0, 0, 1, 1],
+    name: '',
+    description: '',
+    mouseEvent: {},
+    drawing: 0,
+    editing: 0,
+    edit: 0,
+  };
```

Comment:
Here is the application state in all its glory!
The values labeled with single characters represent indices or coordinates.
The "buffer" stores a single waypoint configuration when editing.
The "changed" state indicates whether the "design" has been changed.
The "design" stores the entire configuration as loaded from yaml.
The "w", "g", and "s" are waypoint, group, and story indices respectfully.
The "v" gives the viewport [zoom, x, y] with x and y in normalized [0,1] space over the openseadragon content.
The "o" gives the rectangular overlay in [x0, y0, x1, y1]
The "name" should be a "window.btoa" encoded waypoint name
The "description" should be a "window.btoa" encoded waypoint description
mouseEvent stores the jquery event giving the current mouse position
drawing triggers the start of the region of interest box if 1 or the end of the box if 2
editing opens/closes the editor interface
edit enables/disables opening/closing the editor interface